File: backend/src/linuxRuntime/desktopScanner.py
import os
import glob
import json
import re
import sys

SEARCH_DIRS = [
    '/usr/share/applications',
    '/usr/local/share/applications',
    '/root/.local/share/applications',
    '/var/lib/snapd/desktop/applications',
    '/var/lib/flatpak/exports/share/applications'
]

# Standard icon directories to search for icon assets
ICON_THEME_DIRS = [
    '/usr/share/icons/hicolor',
    '/usr/share/icons/Adwaita',
    '/usr/share/icons/Papirus',
    '/usr/share/icons/breeze',
    '/usr/share/icons/locolor',
    '/usr/share/pixmaps',
]

def clean_exec(raw):
    if not raw:
        return ''
    # Strip field codes: %f, %F, %u, %U, %d, %D, %n, %N, %i, %c, %k, %v, %m
    cleaned = re.sub(r'%[fFuUdDnNickvm]', '', raw).strip()
    return cleaned

def map_category(cats):
    cat_set = {c.lower() for c in cats}
    if cat_set & {'development', 'programming', 'ide', 'debugger', 'building'}:
        return 'development'
    if cat_set & {'network', 'webbrowser', 'email', 'chat', 'feed', 'filetransfer', 'remoteaccess'}:
        return 'internet'
    if cat_set & {'graphics', '2dgraphics', 'rastergraphics', 'vectorgraphics', 'photography', 'viewer', 'image'}:
        return 'graphics'
    if cat_set & {'audiovideo', 'audio', 'video', 'player', 'recorder', 'music', 'midi'}:
        return 'multimedia'
    if cat_set & {'office', 'wordprocessor', 'spreadsheet', 'presentation', 'publishing', 'finance', 'projectmanagement'}:
        return 'office'
    if cat_set & {'security', 'networksecurity', 'forensics', 'kali', 'penetrationtesting', 'vulnerability'}:
        return 'security'
    return 'utilities'

def get_emoji_fallback(category, name):
    name_lower = name.lower()
    if 'terminal' in name_lower or 'uxterm' in name_lower or 'xterm' in name_lower:
        return '🖥️'
    if 'calc' in name_lower:
        return '🧮'
    if 'edit' in name_lower or 'pad' in name_lower or 'vim' in name_lower:
        return '📝'
    if 'browser' in name_lower or 'web' in name_lower:
        return '🌐'
    if 'view' in name_lower or 'image' in name_lower or 'photo' in name_lower:
        return '🎨'
    if 'play' in name_lower or 'media' in name_lower or 'music' in name_lower or 'audio' in name_lower:
        return '🎬'
    if 'parted' in name_lower or 'disk' in name_lower:
        return '💾'
    if 'shark' in name_lower or 'sniff' in name_lower or 'scan' in name_lower:
        return '🦈'

    fallbacks = {
        'development': '💻',
        'internet': '🌐',
        'graphics': '🎨',
        'multimedia': '🎬',
        'office': '📄',
        'security': '🛡️',
        'utilities': '⚙️'
    }
    return fallbacks.get(category, '📦')

def scan_desktop_entries():
    apps = []
    seen_ids = set()

    for d in SEARCH_DIRS:
        if not os.path.exists(d):
            continue
        for path in glob.glob(os.path.join(d, '*.desktop')):
            try:
                with open(path, 'r', encoding='utf-8', errors='ignore') as f:
                    content = f.read()
            except Exception:
                continue

            in_entry = False
            props = {}
            for line in content.splitlines():
                line = line.strip()
                if not line or line.startswith('#'):
                    continue
                if line.startswith('['):
                    in_entry = (line == '[Desktop Entry]')
                    continue
                if in_entry and '=' in line:
                    k, v = line.split('=', 1)
                    k = k.strip()
                    v = v.strip()
                    if k not in props:
                        props[k] = v

            if props.get('Type', 'Application') != 'Application':
                continue
            if props.get('NoDisplay', '').lower() in ('true', '1') or props.get('Hidden', '').lower() in ('true', '1'):
                continue

            raw_exec = props.get('Exec', '')
            if not raw_exec:
                continue

            exec_cmd = clean_exec(raw_exec)
            name = props.get('Name[pt_BR]') or props.get('Name[pt]') or props.get('Name') or os.path.basename(path)[:-8]
            generic_name = props.get('GenericName[pt_BR]') or props.get('GenericName[pt]') or props.get('GenericName') or ''
            comment = props.get('Comment[pt_BR]') or props.get('Comment[pt]') or props.get('Comment') or generic_name or ''
            icon = props.get('Icon', '')
            terminal = props.get('Terminal', '').lower() in ('true', '1')
            cats_raw = props.get('Categories', '')
            cats = [c.strip() for c in cats_raw.split(';') if c.strip()]
            primary_cat = map_category(cats)
            emoji_icon = get_emoji_fallback(primary_cat, name)

            base_id = os.path.basename(path)[:-8].lower()
            if base_id in seen_ids:
                continue
            seen_ids.add(base_id)

            apps.append({
                'id': base_id,
                'name': name,
                'genericName': generic_name,
                'comment': comment,
                'command': exec_cmd,
                'rawExec': raw_exec,
                'icon': icon or emoji_icon,
                'iconName': icon,
                'emojiFallback': emoji_icon,
                'categories': cats,
                'category': primary_cat,
                'terminal': terminal,
                'desktopFile': path,
                'installed': True,
                'isDiscovered': True
            })

    # Sort alphabetically by name
    apps.sort(key=lambda a: a['name'].lower())
    return apps

if __name__ == '__main__':
    result = scan_desktop_entries()
    json.dump(result, sys.stdout, indent=2, ensure_ascii=False)

File: backend/src/linuxRuntime/desktopScanner.js
import fs from 'node:fs';
import path from 'node:path';
import { execFile } from 'node:child_process';
import { promisify } from 'node:util';
import { WSL_EXE, getWslSnapshot, safeChildEnvironment, validateInstalledAsync } from '../wsl/distroService.js';
import { resolveLinuxIconPath } from './iconResolver.js';

const execFileAsync = promisify(execFile);

const SCAN_CACHE_TTL_MS = 15_000;
let cachedDistro = null;
let lastScanTime = 0;
let cachedDiscoveredApps = [];

const SEARCH_SUBDIRS = [
  'usr/share/applications',
  'usr/local/share/applications',
  'root/.local/share/applications',
  'var/lib/snapd/desktop/applications',
  'var/lib/flatpak/exports/share/applications'
];

function cleanExec(raw) {
  if (!raw) return '';
  // Strip field codes: %f, %F, %u, %U, %d, %D, %n, %N, %i, %c, %k, %v, %m
  return String(raw).replace(/%[fFuUdDnNickvm]/g, '').trim();
}

function mapCategory(cats) {
  const catSet = new Set(cats.map(c => c.toLowerCase()));
  if ([...catSet].some(c => ['development', 'programming', 'ide', 'debugger', 'building'].includes(c))) {
    return 'development';
  }
  if ([...catSet].some(c => ['network', 'webbrowser', 'email', 'chat', 'feed', 'filetransfer', 'remoteaccess'].includes(c))) {
    return 'internet';
  }
  if ([...catSet].some(c => ['graphics', '2dgraphics', 'rastergraphics', 'vectorgraphics', 'photography', 'viewer', 'image'].includes(c))) {
    return 'graphics';
  }
  if ([...catSet].some(c => ['audiovideo', 'audio', 'video', 'player', 'recorder', 'music', 'midi'].includes(c))) {
    return 'multimedia';
  }
  if ([...catSet].some(c => ['office', 'wordprocessor', 'spreadsheet', 'presentation', 'publishing', 'finance', 'projectmanagement'].includes(c))) {
    return 'office';
  }
  if ([...catSet].some(c => ['security', 'networksecurity', 'forensics', 'kali', 'penetrationtesting', 'vulnerability'].includes(c))) {
    return 'security';
  }
  return 'utilities';
}

function getEmojiFallback(category, name) {
  const nameLower = String(name || '').toLowerCase();
  if (nameLower.includes('terminal') || nameLower.includes('uxterm') || nameLower.includes('xterm')) return '🖥️';
  if (nameLower.includes('calc')) return '🧮';
  if (nameLower.includes('edit') || nameLower.includes('pad') || nameLower.includes('vim') || nameLower.includes('note')) return '📝';
  if (nameLower.includes('browser') || nameLower.includes('web') || nameLower.includes('fox') || nameLower.includes('chrome')) return '🌐';
  if (nameLower.includes('view') || nameLower.includes('image') || nameLower.includes('photo') || nameLower.includes('paint') || nameLower.includes('gimp')) return '🎨';
  if (nameLower.includes('play') || nameLower.includes('media') || nameLower.includes('music') || nameLower.includes('audio') || nameLower.includes('vlc')) return '🎬';
  if (nameLower.includes('parted') || nameLower.includes('disk') || nameLower.includes('file')) return '📁';
  if (nameLower.includes('shark') || nameLower.includes('sniff') || nameLower.includes('scan') || nameLower.includes('nmap')) return '🛡️';

  const fallbacks = {
    development: '💻',
    internet: '🌐',
    graphics: '🎨',
    multimedia: '🎬',
    office: '📄',
    security: '🛡️',
    utilities: '⚙️'
  };
  return fallbacks[category] || '📦';
}

function parseDesktopFileContent(content, filename) {
  let inEntry = false;
  const props = {};
  for (const rawLine of content.split(/\r?\n/)) {
    const line = rawLine.trim();
    if (!line || line.startsWith('#')) continue;
    if (line.startsWith('[')) {
      inEntry = (line === '[Desktop Entry]');
      continue;
    }
    if (inEntry && line.includes('=')) {
      const idx = line.indexOf('=');
      const k = line.slice(0, idx).trim();
      const v = line.slice(idx + 1).trim();
      if (!props[k]) props[k] = v;
    }
  }

  if (props.Type && props.Type !== 'Application') return null;
  if (['true', '1'].includes(String(props.NoDisplay || '').toLowerCase())) return null;
  if (['true', '1'].includes(String(props.Hidden || '').toLowerCase())) return null;

  const rawExec = props.Exec || '';
  if (!rawExec) return null;

  const baseId = path.basename(filename, '.desktop').toLowerCase();
  const name = props['Name[pt_BR]'] || props['Name[pt]'] || props.Name || baseId;
  const genericName = props['GenericName[pt_BR]'] || props['GenericName[pt]'] || props.GenericName || '';
  const comment = props['Comment[pt_BR]'] || props['Comment[pt]'] || props.Comment || genericName || '';
  const icon = props.Icon || '';
  const terminal = ['true', '1'].includes(String(props.Terminal || '').toLowerCase());
  const catsRaw = props.Categories || '';
  const categories = catsRaw.split(';').map(c => c.trim()).filter(Boolean);
  const category = mapCategory(categories);
  const emojiFallback = getEmojiFallback(category, name);
  const command = cleanExec(rawExec);

  return {
    id: baseId,
    name,
    genericName,
    comment,
    command,
    rawExec,
    icon: icon || emojiFallback,
    iconName: icon,
    emojiFallback,
    categories,
    category,
    terminal,
    desktopFile: filename,
    installed: true,
    isDiscovered: true
  };
}

export async function scanDiscoveredLinuxApps(requestedDistro, { force = false } = {}) {
  const snapshot = await getWslSnapshot();
  const distro = typeof requestedDistro === 'string' && requestedDistro.trim()
    ? requestedDistro.trim()
    : snapshot.preferred || snapshot.default || 'kali-linux';

  if (!force && cachedDistro === distro && Date.now() - lastScanTime < SCAN_CACHE_TTL_MS && cachedDiscoveredApps.length > 0) {
    return cachedDiscoveredApps;
  }

  const baseWsl = `\\\\wsl.localhost\\${distro}`;
  const isDirectAccessible = fs.existsSync(baseWsl);

  const apps = [];
  const seenIds = new Set();

  if (isDirectAccessible) {
    for (const sub of SEARCH_SUBDIRS) {
      const winDir = path.join(baseWsl, sub.replace(/\//g, '\\'));
      if (!fs.existsSync(winDir)) continue;
      try {
        const files = fs.readdirSync(winDir);
        for (const file of files) {
          if (!file.endsWith('.desktop')) continue;
          const fullPath = path.join(winDir, file);
          try {
            const content = fs.readFileSync(fullPath, 'utf8');
            const linuxPath = `/${sub}/${file}`;
            const parsed = parseDesktopFileContent(content, linuxPath);
            if (parsed && !seenIds.has(parsed.id)) {
              seenIds.add(parsed.id);
              // Check if icon exists and formulate iconUrl
              const iconResolved = resolveLinuxIconPath(distro, parsed.iconName);
              if (iconResolved) {
                parsed.iconUrl = `/__cloudos/linux-runtime/icons/${encodeURIComponent(parsed.id)}?distro=${encodeURIComponent(distro)}`;
              } else {
                parsed.iconUrl = null;
              }
              apps.push(parsed);
            }
          } catch {}
        }
      } catch {}
    }
  } else {
    // Fallback: Run scanner script inside WSL
    try {
      const scannerPyPath = path.join(path.dirname(new URL(import.meta.url).pathname.replace(/^\/([a-zA-Z]:)/, '$1')), 'desktopScanner.py');
      const { stdout } = await execFileAsync(WSL_EXE, [
        '-d', distro,
        '--exec', 'python3', '-c',
        fs.readFileSync(scannerPyPath, 'utf8')
      ], {
        encoding: 'utf8',
        env: safeChildEnvironment(),
        timeout: 10_000,
        windowsHide: true,
        maxBuffer: 4 * 1024 * 1024
      });

      const parsedList = JSON.parse(stdout || '[]');
      for (const item of parsedList) {
        if (!seenIds.has(item.id)) {
          seenIds.add(item.id);
          apps.push(item);
        }
      }
    } catch (err) {
      console.warn('⚠️ [DesktopScanner] Erro no fallback WSL:', err.message);
    }
  }

  // Sort alphabetically
  apps.sort((a, b) => a.name.localeCompare(b.name, 'pt-BR'));

  cachedDistro = distro;
  lastScanTime = Date.now();
  cachedDiscoveredApps = apps;
  return apps;
}

export function invalidateDiscoveryCache() {
  lastScanTime = 0;
  cachedDiscoveredApps = [];
}
